clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1087,23 +1076,38 @@ private Record nextFetchedRecord() {
         }
 
         private List<ConsumerRecord<K, V>> fetchRecords(int maxRecords) {
+            // Error when fetching the next record before deserialization.
+            if (cachedRecordFetchException != null && lastRecord == null)
+                throw cachedRecordFetchException;
+
             if (isFetched)
                 return Collections.emptyList();
 
             List<ConsumerRecord<K, V>> records = new ArrayList<>();
             try {
                 for (int i = 0; i < maxRecords; i++) {
-                    Record record = nextFetchedRecord();
+                    Record record = lastRecord;
+                    // Only move to next record if there was no exception in the last fetch. Otherwise we should
+                    // use the last record to do deserialization again.
+                    if (cachedRecordFetchException == null) {
+                        lastRecord = null;
+                        record = nextFetchedRecord();
+                        lastRecord = record;
+                    }
                     if (record == null)
                         break;
-
                     records.add(parseRecord(partition, currentBatch, record));
                     recordsRead++;
                     bytesRead += record.sizeInBytes();
                     nextFetchOffset = record.offset() + 1;
+                    // In some cases, the deserialization may have thrown an exception and the retry may succeed,
+                    // we allow user to move forward in this case.
+                    cachedRecordFetchException = null;
                 }
-            } catch (KafkaException e) {
-                hasExceptionInLastFetch = true;
+            } catch (Exception e) {
+                cachedRecordFetchException = new KafkaException("Received exception when fetching the next record. "
```

Comment:
Would it make sense to cache the underlying exception instead and throw a new `KafkaException` with this message every time we hit that case above? 

Also, one thing I realized is that the user doesn't have any way to know which topic partition contained the fetch error, so I'm not sure how they should seek past the current offset. Maybe we need a new exception type (not necessarily in this patch) which has a reference to the topic partition and offset?